scripts/specCoverageReport.ts

```diff
@@ -0,0 +1,424 @@
+  public parentNode: SpecItemNode | null = null;
+  public childNodes: SpecItemNode[] = [];
+  public specCoverages: TestSpecCoverage[] = [];
+  private _coveragePercentage: number | null = null;
+
+  constructor(specItem: SpecItem) {
+    this.level = specItem.level;
+    this.key = specItem.key;
+    this.description = specItem.description;
+  }
+
+  public isRoot(): boolean {
+    return this.parentNode == null;
+  }
+
+  public isLeaf(): boolean {
+    return this.childNodes.length === 0;
+  }
+
+  public getCoveragePercentage(): number {
+    // general rules for coverage:
+    // if the spec item is covered by the test with @spec tag - it contributes a weigth of 1 (100%).
+    // if the spec item is covered by the test with @specpartial tag - it contributes a weigth of 0.5 (50%).
+    // otherwise it contributes a weigth of 0 (0%).
+    // when multiple nodes are contributing to coverage calculation (like for parent nodes),
+    // then their coverage percentages are added and then divided by the number of nodes to calculate the average coverage for the node.
+
+    // memoize coverage percentage
+    if (this._coveragePercentage != null) {
+      return this._coveragePercentage;
+    }
+
+    // leaf nodes calculate coverage percentage only based on their own spec coverage
+    if (this.isLeaf()) {
+      this._coveragePercentage = this._getCoveragePercentageForLeaf();
+      return this._coveragePercentage;
+    }
+
+    // parent nodes calculate coverage percentage based on their own and children's coverages
+    this._coveragePercentage = this._getCoveragePercentageForParent();
+    return this._coveragePercentage;
+  }
+
+  public getLinkToSpecItem(): string {
+    return `https://sdk.ably.com/builds/ably/specification/main/features/#${this.key}`;
+  }
+
+  public toSpecItemView(): SpecItem {
+    return {
+      level: this.level,
+      key: this.key,
+      description: this.description,
+    };
+  }
+
+  public toStringView(): string {
+    let specItemLine = '';
+
+    specItemLine += new Array(this.level - 1).fill('    ').join('');
+    specItemLine += `(${this.key})`;
+    specItemLine += ` ${this.description}`;
+
+    return specItemLine;
+  }
+
+  public toJSON(): object {
+    return {
+      ...this.toSpecItemView(),
+      parentNode: this.parentNode?.toSpecItemView(),
+      childNodes: this.childNodes,
+      specCoverages: this.specCoverages,
+    };
+  }
+
+  private _getCoveragePercentageForLeaf(): number {
+    if (this.specCoverages.findIndex((x) => x.coverageType === SpecCoverageType.spec) !== -1) {
+      return 1;
+    }
+
+    if (this.specCoverages.findIndex((x) => x.coverageType === SpecCoverageType.specpartial) !== -1) {
+      return 0.5;
+    }
+
+    return 0;
+  }
+
+  private _getCoveragePercentageForParent(): number {
+    // here we use the next assumption:
+    // if all children spec items are covered by the test (average coverage is 1), then we can consider the parent covered too,
```

Comment:
So, we're using information about the ably-js test suite (i.e. whether a given spec point has a test) to infer information about the spec (i.e. whether a given spec point needs a test)? I’m not sure that’s a good idea; it feels like the spec should be the source of truth.

I think that it'd be better if we could supply some metadata alongside the spec indicating which spec points are describing a testable behaviour, and which are either purely informational or purely a heading for their sub-points. Until then, I’d suggest that we perhaps not try and infer whether a given spec point needs to be explicitly tested or not.

Reply:
> I’m not sure that’s a good idea; it feels like the spec should be the source of truth.

Definitely agree. I was trying to work with the fact that some spec item parents in the spec are only described and can be tested by their children spec items, but I can see the shortcomings of this approach. Test suite should no do such assumptions.

> I think that it'd be better if we could supply some metadata alongside the spec indicating which spec points are describing a testable behaviour, and which are either purely informational or purely a heading for their sub-points.

Yep, same idea was floating around regarding future improvements to the spec.
As you've suggested, until then, I'll change the report to expect that every spec item is currently needs to be tested, as there is no other indication available atm.

Reply:
Updated in https://github.com/ably/ably-js/pull/1811/commits/b1ab4b0c64fcec47cc5c798e498be17036d0ce7c